### PR TITLE
feat(manager/pip-compile): support basic arguments

### PR DESCRIPTION
**Pull request by msw-kialo**
## Changes:

Parse `requirements.txt` file to detect previously used arguments to
call `pip-compile`. Most notably are `--generated-hashes` to support the [Hash-Checking Mode](https://pip-tools.readthedocs.io/en/latest/#using-hashes), and `--allow-unsafe` to [prepare for the planned change](https://pip-tools.readthedocs.io/en/latest/#deprecations).
Furthermore, `--output-file` is supported (but with Renovate's value) as it is commonly used (like by Dependabot).

## Context:

Contribution to #2334

## Open Questions

I wrote this as a minimal viable version (at least for my use-case) to get the discussion starting. I was unsure regarding the following points and would be happy to get comments in the review if needed:

- Better approaches to parse the file (e.g. parsing line based instead of this single regex)?
- How much logging is wanted for this feature (I think it would be good to have a way to figure how why arguments were rejected)?
- Does this need more/higher unit tests?
- Should I allow as many options as possible (like going through the list of currently supported arguments). I haven't used most of them and can't test whether it actually works.
- Some arguments could also be detected/guest by inspecting the `requirements.txt` content itself (like `--hash=` arguments should imply `--generate-hashes` this should also work if `--no-header` or a custom wrapper command is used).
- How much documentation is wanted here?

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/pip-compile/artifacts.spec.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { exec as _exec } from 'child_process';
 import _fs from 'fs-extra';
 import { join } from 'upath';
 import { envMock, mockExecAll } from '../../../test/exec-util';
-import { git, mocked } from '../../../test/util';
+import { git, loadFixture, mocked } from '../../../test/util';
```
Comment (viceice) [inline lib/manager/pip-compile/artifacts.spec.ts:5]: Please use `Fixtures.get`  from `test/fixtures`

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/manager/pip-compile/artifacts.ts [hunk 2]
```diff
@@ -40,6 +40,40 @@ function getPipToolsConstraint(config: UpdateArtifactsConfig): string {
...
+  const headers = constraintLineRegex.exec(content);
+  const args = ['pip-compile'];
+  if (headers) {
+    logger.debug({ header: headers[0] }, 'Found pip-compile header');
+    for (const argument of split(headers.groups.arguments)) {
+      if (['--allow-unsafe', '--generate-hashes'].includes(argument)) {
+        args.push(argument);
+      } else if (argument.startsWith('--output-file=')) {
+        // we don't trust the user-supplied output-file argument; use our value here
+        args.push(`--output-file=${upath.parse(outputFileName).base}`);
```
Comment (viceice) [inline lib/manager/pip-compile/artifacts.ts:61]: I think we should warn if there is a mismatch. We should at least issue an info log message.

File: lib/manager/pip-compile/artifacts.ts [hunk 3]
```diff
@@ -40,6 +40,40 @@ function getPipToolsConstraint(config: UpdateArtifactsConfig): string {
...
+      } else if (argument.startsWith('--')) {
+        logger.trace(
+          { argument: argument },
```
Comment (viceice) [inline lib/manager/pip-compile/artifacts.ts:64]:
```suggestion
          { argument },
```
Simplify. 😉

File: lib/manager/pip-compile/artifacts.ts [hunk 4]
```diff
@@ -40,6 +40,46 @@ function getPipToolsConstraint(config: UpdateArtifactsConfig): string {
...
+        args.push(argument);
+      } else if (argument.startsWith('--output-file=')) {
+        if (argument !== `--output-file=${upath.parse(outputFileName).base}`) {
+          // we don't trust the user-supplied output-file argument; use our value here
+          logger.warn(
+            { argument },
+            'pip-compile was previously executed with an unexpected `--output-file` filename'
+          );
+        }
+        args.push(`--output-file=${upath.parse(outputFileName).base}`);
```
Comment (viceice) [inline lib/manager/pip-compile/artifacts.ts:67]:
```suggestion
        const file = upath.parse(outputFileName).base;
        if (argument !== `--output-file=${file}`) {
          // we don't trust the user-supplied output-file argument; use our value here
          logger.warn(
            { argument },
            'pip-compile was previously executed with an unexpected `--output-file` filename'
          );
        }
        args.push(`--output-file=${file}`);
```

**Review by viceice — COMMENTED** [review]
I still can't apply my suggestions nor update the branch 🤷‍♂️

**Comment by viceice** [comment]
Please enable maintainer access checkbox.

**Review by viceice — APPROVED** [review]

**Pull request merged**